### PR TITLE
APC & Machinery Refactor + Local Powernets

### PR DESCRIPTION
**Pull request by Sirryan2002**
## What Does This PR Do
First step in #19786

- Refactors APCs (Moving stuff to defines, splitting apc.dm into like 4 different files)
- Moves power processing in areas to a local powernet datum as opposed to on the `/area` or `/machinery/power/apc` level
- All machinery will now use a passive (and more CPU performative) method for power consumption where it used to consume power every process interval
- Refactors power_change() to only allow icon/lighting updating when the machine's NOPOWER flag changes
- Standardizes power proc/var usage across machines

Note that this does not rebalance power values nor does it actually have any player facing changes

## Why It's Good For The Game
Power code is mostly baycode. Furthermore it's only slightly less magical and confusing than atmos code so it needs to be brought up to modern standards before I even think about adding/modifying power features (I learned my lesson from econ rework)

## Testing
- Compiled and Ran
- Check, flicked, and played with all buttons and options on APCs
- Depowered the entire station and repowered it
- Fully constructed and deconstructed APCs
- Set up Supermatter Engine and ensured power was flowing correct into SMES and then being delivered to APCs
- Checked to ensure that APCs and light switches can be built properly in newly blueprinted areas
- ran on personal server for an hour to observe possible long-term power issues

no player facing changes so no changelog

### CODE REVIEW

File: code/game/machinery/teleporter.dm [hunk 1]
```diff
@@ -493,6 +491,8 @@
 
 /obj/machinery/teleport/perma/power_change()
 	..()
+	if(!..())
+		return
```
Comment (SteelSlayer) [inline code/game/machinery/teleporter.dm:495]: Double parent call here.

File: code/game/objects/structures/electricchair.dm [hunk 2]
```diff
@@ -91,5 +91,5 @@
 			to_chat(buckled_mob, "<span class='danger'>You feel a deep shock course through your body!</span>")
 			spawn(1)
 				buckled_mob.electrocute_act(110, src, 1)
-	A.power_light = light
+	A.powernet.lighting_powered = light
```
Comment (SteelSlayer) [inline code/game/objects/structures/electricchair.dm:94]: Shouldn't need to declare `var/light` if you do this:
```suggestion
	A.powernet.lighting_powered = A.powernet.has_power(PW_CHANNEL_LIGHTING)
```

File: code/modules/mob/living/silicon/ai/life.dm [hunk 3]
```diff
@@ -141,7 +141,7 @@
 /mob/living/silicon/ai/proc/lacks_power()
 	var/turf/T = get_turf(src)
 	var/area/A = get_area(src)
-	return ((!A.power_equip) && A.requires_power == 1 || isspaceturf(T)) && !isitem(src.loc)
+	return ((!A.powernet.equipment_powered) && A.requires_power == 1 || isspaceturf(T)) && !isitem(src.loc)
```
Comment (SteelSlayer) [inline code/modules/mob/living/silicon/ai/life.dm:144]:
```suggestion
	return (!A.powernet.equipment_powered && A.requires_power == 1 || isspaceturf(T)) && !isitem(src.loc)
```

File: code/modules/power/apc/apc_construction.dm [hunk 4]
```diff
@@ -0,0 +1,135 @@
+
+
+/obj/machinery/power/apc/proc/has_electronics()
+	return electronics_state != APC_ELECTRONICS_NONE
+
+/obj/machinery/power/apc/deconstruct(disassembled = TRUE)
+	if(!(flags & NODECONSTRUCT))
+		if(!(stat & BROKEN))
+			set_broken()
+		if(opened != 2)
```
Comment (SteelSlayer) [inline code/modules/power/apc/apc_construction.dm:10]:
```suggestion
		if(opened != APC_COVER_OFF)
```

File: code/modules/power/apc/apc.dm [hunk 5]
```diff
@@ -0,0 +1,1064 @@
...
+	/*** APC Status Vars ***/
+	/// The wire panel associated with this APC
+	var/datum/wires/apc/wires = null
+	/// Can the APC recieve/transmit power? Determined by the condition of the 2 Main Power Wires
+	var/shorted = FALSE
+	/// Is the APC on and transmitting power to enabled breakers? Think of this var as the master breaker for the APC
+	var/operating = TRUE
+	/// The current charging mode of the APC: not charging, charging, fully charged
+	var/charging = APC_NOT_CHARGING
+	var/longtermpower = 10
```
Comment (SteelSlayer) [inline code/modules/power/apc/apc.dm:88]: A comment for this var would be nice too ❤️ 

Also, probably should be snake case.

File: code/modules/power/lighting.dm [hunk 6]
```diff
@@ -567,13 +567,12 @@
 // if a light is turned off, it won't activate emergency power
 /obj/machinery/light/proc/turned_off()
 	var/area/A = get_area(src)
-	return !A.lightswitch && A.power_light
+	return !A.lightswitch && A.powernet.lighting_powered
```
Comment (SteelSlayer) [inline code/modules/power/lighting.dm:570]: Could you do `!machine_area.lightswitch && machine_area.powernet.lighting_powered` here?

File: code/modules/power/singularity/particle_accelerator/particle_control.dm [hunk 7]
```diff
@@ -150,9 +150,9 @@
 	..()
 	if(stat & NOPOWER)
 		active = FALSE
-		use_power = NO_POWER_USE
+		change_power_mode(NO_POWER_USE)
```
Comment (SteelSlayer) [inline code/modules/power/singularity/particle_accelerator/particle_control.dm:153]: This won't do anything since `NO_POWER_USE` is `0` and the proc immediately returns if `!use_type`. There are several other instances of this too.

File: code/modules/station_goals/bsa.dm [hunk 8]
```diff
@@ -202,7 +202,10 @@
 			top_layer.layer = 4.1
 			icon_state = "cannon_east"
 	overlays += top_layer
-	reload()
+
+/obj/machinery/bsa/full/Initialize(mapload)
+	. = ..()
+	addtimer(CALLBACK(src, PROC_REF(reload)), 30 SECONDS) //wait until shit is loaded before we start using power to reload!
```
Comment (SteelSlayer) [inline code/modules/station_goals/bsa.dm:208]: Does `LateInitialize` work or are things still not loaded in that time?

File: code/game/area/areas_power.dm [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+			if(state)
+				SSalarm.cancelAlarm("Power", src, source)
+			else
+				SSalarm.triggerAlarm("Power", src, cameras, source)
+
+
+/**
+  * Called when the area power status changes
+  *
+  * Updates the area icon, calls power change on all machines in the area, and sends the `COMSIG_AREA_POWER_CHANGE` signal.
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/game/area/areas_power.dm:29]: good catch

File: code/game/machinery/machinery.dm [hunk 10]
```diff
@@ -94,32 +106,71 @@
...
+		if(IDLE_POWER_USE)
+			remove_static_power(power_channel, idle_power_consumption)
+		if(ACTIVE_POWER_USE)
+			remove_static_power(power_channel, active_power_consumption)
+
+	switch(use_type)
+		if(IDLE_POWER_USE)
+			add_static_power(power_channel, idle_power_consumption)
+		if(ACTIVE_POWER_USE)
+			add_static_power(power_channel, active_power_consumption)
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/game/machinery/machinery.dm:165]: Yes, there's probably a more elegant way to do this. Since static power is only tracked on the powernet, if I need to alter the static power this machine uses from `idle` to `active` or to `0` power consumption then I need to take into account how static power it is using now and then add/subtract the difference, in this case I figured it was just easier to remove the static usage entirely first and then add the new one.

File: code/modules/power/singularity/particle_accelerator/particle_control.dm (hunk 7 above)
Comment (Sirryan2002) [reply code/modules/power/singularity/particle_accelerator/particle_control.dm:153]: that needs to be a `isnull(use_type)` check, oops. Thank you for catching that.

File: code/modules/station_goals/bsa.dm (hunk 8 above)
Comment (Sirryan2002) [reply code/modules/station_goals/bsa.dm:208]: LateInitialize might work here actually

File: code/game/area/areas_power.dm [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  */
+/area/proc/power_change()
+	update_icon(UPDATE_ICON_STATE)
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/game/area/areas_power.dm:32]: I'm not sure, this is leftover code, I'll check with vi3trice or something they'd probably know

File: code/game/area/areas_power.dm [hunk 12]
```diff
@@ -0,0 +1,32 @@
+///**ALL POWER RELATED PROCS AND CODE DEFINED ON THE /area LEVEL**///
+
+/**
+  * Generate a power alert for this area
+  *
+  * Sends to all ai players, alert consoles, drones and alarm monitor programs in the world
+  */
+/area/proc/poweralert(state, obj/source)
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/game/area/areas_power.dm:8]: done

File: code/game/area/areas_power.dm (hunk 9 above)
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/game/area/areas_power.dm]: edited

File: code/game/machinery/flasher.dm [hunk 13]
```diff
@@ -72,7 +71,7 @@
 		return flash()
 
 /obj/machinery/flasher/proc/flash()
-	if(!(powered()))
+	if(!(has_power()))
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/game/machinery/flasher.dm]: done

File: code/game/machinery/teleporter.dm (hunk 1 above)
Comment (Sirryan2002) [reply code/game/machinery/teleporter.dm]: extra parent call removed

File: code/modules/mob/living/silicon/ai/life.dm (hunk 3 above)
Comment (Sirryan2002) [reply code/modules/mob/living/silicon/ai/life.dm]: done

File: code/modules/power/apc/apc_construction.dm (hunk 4 above)
Comment (Sirryan2002) [reply code/modules/power/apc/apc_construction.dm]: done

File: code/modules/power/apc/apc.dm (hunk 5 above)
Comment (Sirryan2002) [reply code/modules/power/apc/apc.dm:89]: comment added, will try and get snake case done as well

File: code/modules/power/singularity/particle_accelerator/particle_control.dm (hunk 7 above)
Comment (Sirryan2002) [reply code/modules/power/singularity/particle_accelerator/particle_control.dm:153]: fixed

File: code/modules/station_goals/bsa.dm (hunk 8 above)
Comment (Sirryan2002) [reply code/modules/station_goals/bsa.dm]: moved to late initialize

File: code/game/objects/structures/electricchair.dm (hunk 2 above)
Comment (Sirryan2002) [reply code/game/objects/structures/electricchair.dm]: this entire section has been removed, I'm not sure why the electric chair was setting power channels, that's whack

File: code/modules/power/lighting.dm (hunk 6 above)
Comment (Sirryan2002) [reply code/modules/power/lighting.dm]: done

**Review by S34NW — COMMENTED** [review]
Mapping changes are fine

**Comment by Sirryan2002** [comment]
will deconflict this in a moment

File: code/game/machinery/portable_turret.dm [hunk 14]
```diff
@@ -1022,7 +1019,10 @@ GLOBAL_LIST_EMPTY(turret_icons)
 	syndicate = TRUE
 	installation = null
 	always_up = TRUE
-	use_power = NO_POWER_USE
+	requires_power = FALSE
+	power_state = NO_POWER_USE
+	idle_power_consumption = 0
+	active_power_consumption = 0
```
Comment (Henri215) [inline code/game/machinery/portable_turret.dm:1025]:
```suggestion
```
Machinery already sets both as 0, the power_state being no_power_use should be enough to indicate that it doesnt use power. Same happens at beacon.dm and atmospheric.dm

**Comment by Sirryan2002** [comment]
the tests failing aren't due to the PR itself :)

File: code/game/area/areas_power.dm (hunk 11 above)
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/game/area/areas_power.dm:34]: yeah, this is getting snapped 💀

File: code/modules/atmospherics/machinery/atmospherics.dm [hunk 15]
```diff
@@ -14,9 +14,10 @@ Pipelines + Other Objects -> Pipe network
 	resistance_flags = FIRE_PROOF
 	max_integrity = 200
 	plane = GAME_PLANE
-	idle_power_usage = 0
-	active_power_usage = 0
-	power_channel = ENVIRON
+	power_state = NO_POWER_USE
+	idle_power_consumption = 0
+	active_power_consumption = 0
+	power_channel = PW_CHANNEL_ENVIRONMENT
```
Comment (Henri215) [inline code/modules/atmospherics/machinery/atmospherics.dm:20]:
```suggestion
	power_state = NO_POWER_USE
	power_channel = PW_CHANNEL_ENVIRONMENT
```

File: code/game/machinery/Beacon.dm [hunk 16]
```diff
@@ -7,8 +7,7 @@
 	layer = WIRE_LAYER
 	plane = FLOOR_PLANE
 	anchored = TRUE
-	use_power = IDLE_POWER_USE
-	idle_power_usage = 0
+	idle_power_consumption = 0
 	var/syndicate = FALSE
```
Comment (Henri215) [inline code/game/machinery/Beacon.dm:11]:
```suggestion
	var/syndicate = FALSE
```

**Review by Henri215 — COMMENTED** [review]
From the beacon.dm and atmospheric.dm

File: code/game/area/areas.dm [hunk 17]
```diff
@@ -18,21 +18,16 @@
...
-	var/used_light = FALSE
-	var/used_environ = FALSE
-	var/static_equip
-	var/static_light = FALSE
-	var/static_environ
+	var/always_unpowered = FALSE
 
-	var/has_gravity = TRUE
+	var/datum/local_powernet/powernet = null
 	var/list/apc = list()
```
Comment (hal9000PR) [inline code/game/area/areas.dm:27]: autodoc these please

File: code/game/area/areas.dm [hunk 18]
```diff
@@ -118,9 +123,10 @@
 
 	return INITIALIZE_HINT_LATELOAD
 
-/area/LateInitialize()
-	. = ..()
-	power_change()		// all machines set to current power level, also updates lighting icon
+/area/proc/create_powernet()
+	powernet = new()
+	powernet.powernet_area = src
+	return powernet
```
Comment (hal9000PR) [inline code/game/area/areas.dm:129]: I think this proc should contain the code from lines 94 to 105. because this *does* create the powernet but not entirely, half of the powernet is setup in the initialise proc.

What happens if this is called on an area with a powernet already? is there ever a need to replace a powernet? maybe if we ever have a powernet subtype?

File: code/game/area/areas_power.dm [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+	if(state == poweralm)
+		return
+	poweralm = state
+	if(!istype(source))	//Only report power alarms on the z-level where the source is located.
+		return
```
Comment (hal9000PR) [inline code/game/area/areas_power.dm:13]: what does this have to do with Zlevles?

**Review by hal9000PR — CHANGES_REQUESTED** [review]
few pre-TM comments

File: code/game/area/areas_power.dm (hunk 19 above)
Comment (Sirryan2002) [reply code/game/area/areas_power.dm:13]: cleaned up in next commit, this is probably just leftover stuff

File: code/__DEFINES/power.dm [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+
+//Power use
+#define NO_POWER_USE 0
+#define IDLE_POWER_USE 1
+#define ACTIVE_POWER_USE 2
+
+//APC charging
+#define APC_NOT_CHARGING 0
+#define APC_IS_CHARGING 1
+#define APC_FULLY_CHARGED 2
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/__DEFINES/power.dm:25]: I'll turn the machine stat ones into bit shifts, apc charging & power use aren't flags however

File: code/game/area/areas.dm (hunk 18 above)
Comment (Sirryan2002) [reply code/game/area/areas.dm]: done

File: code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm [hunk 21]
```diff
@@ -56,32 +56,29 @@
 	if(welded)
 		. += "It seems welded shut."
 
-/obj/machinery/atmospherics/unary/vent_scrubber/auto_use_power()
-	if(!powered(power_channel))
+/obj/machinery/atmospherics/unary/vent_scrubber/proc/adjust_power_consumption()
```
Comment (SteelSlayer) [inline code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm:59]: This proc isn't used anywhere.

File: code/modules/power/powernets/local_powernet.dm [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+  *
+  * Machine/Turf/Item -> Local Powernet -> APC -> Terminal -> Wirenet
+  *
+  * Non-machines cannot register to the local powernet and should instead listen for power change signals
+*/
+/datum/local_powernet
+	/// The area this local powernet is attached to
+	var/area/powernet_area = null
+	/// The APC associated with this powernet
+	var/obj/machinery/power/apc/powernet_apc = null
```
Comment (SteelSlayer) [inline code/modules/power/powernets/local_powernet.dm:15]: Make sure to clear this ref in `apc/Destroy`. Or do signal shenanigans.

File: code/modules/power/apc/apc.dm [hunk 23]
```diff
@@ -0,0 +1,1063 @@
...
+	var/update_state = -1
+	var/update_overlay = -1
+	var/global/status_overlays = FALSE
+	var/updating_icon = FALSE
+	var/global/list/status_overlays_lock
+	var/global/list/status_overlays_charging
+	var/global/list/status_overlays_equipment
+	var/global/list/status_overlays_lighting
+	var/global/list/status_overlays_environ
+	var/keep_preset_name = FALSE
```
Comment (SteelSlayer) [inline code/modules/power/apc/apc.dm:145]: What the hell is this disgusting crap? APC code is wild.

**Review by SteelSlayer — COMMENTED** [review]
Small stuff. Should be good for a TM.

File: code/game/area/areas.dm [hunk 24]
```diff
@@ -13,26 +13,26 @@
...
-	var/valid_territory = TRUE //used for cult summoning areas on station zlevel
-	var/map_name // Set in New(); preserves the name set by the map maker, even if renamed by the Blueprints.
-	var/lightswitch = TRUE
 
-	var/debug = FALSE
+	/// used for cult summoning areas on station zlevel
+	var/valid_territory = TRUE
+	/// Set in New(); preserves the name set by the map maker, even if renamed by the Blueprints.
+	var/map_name
+	/// Is the lightswitching in this area on? Controls weather or not lights are on and off
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/game/area/areas.dm:21]: fixed

File: code/game/machinery/machinery.dm [hunk 25]
```diff
@@ -9,38 +8,49 @@
...
-		//2 = run auto, use active
-	var/idle_power_usage = 0
-	var/active_power_usage = 0
-	var/power_channel = EQUIP //EQUIP,ENVIRON or LIGHT
+
+	/// How is this machine currently passively consuming power?
+	var/power_state = IDLE_POWER_USE
+	/// Does this machine require power?
+	var/requires_power = TRUE
+	/// How much power does this machine consume when it is idleing
```
[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/game/machinery/machinery.dm:16]: fixed

File: code/modules/power/apc/apc.dm (hunk 23 above)
Comment (Sirryan2002) [reply code/modules/power/apc/apc.dm:145]: yeah... it's pretty useful but its so hard to figure out. It's all just performance-saving stuff for overlays.

File: code/modules/power/powernets/local_powernet.dm (hunk 22 above)
Comment (Sirryan2002) [reply code/modules/power/powernets/local_powernet.dm:15]: I believed it's already handled in apc destroy

File: code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm (hunk 21 above)
Comment (Sirryan2002) [reply code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm:59]: I just removed it

**Review by hal9000PR — APPROVED** [review]
